File: common/src/main/java/com/proximyst/tab/platform/IPlatform.java
package com.proximyst.tab.platform;

import com.proximyst.tab.platform.factories.IPlayerFactory;
import java.util.Collection;
import java.util.UUID;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

/**
 * A platform for the plugin to run on.
 *
 * @param <Player> The type of player for this platform.
 */
public interface IPlatform<Player extends IPlayer> {
  /**
   * @return Whether this platform is a proxy.
   */
  default boolean isProxy() {
    return false;
  }

  /**
   * @return All currently connected players.
   */
  @NotNull
  Collection<? extends Player> getConnectedPlayers();

  /**
   * Get a connected player by their UUID.
   *
   * @param uuid The UUID to get.
   * @return The player found or null if none.
   */
  @Nullable
  Player getPlayer(@NotNull UUID uuid);

  /**
   * Get a connected player by their name.
   *
   * @param name The name to get.
   * @return The player found or null if none.
   */
  @Nullable
  Player getPlayer(@NotNull String name);

  /**
   * @return This platform's {@link IPlayerFactory}.
   */
  @NotNull
  IPlayerFactory<?, Player> getPlayerFactory();
}

File: common/src/main/kotlin/com/proximyst/tab/common/IPlatform.kt
package com.proximyst.tab.common

import java.util.*

/**
 * A platform on which the plugin can run and retrieve data.
 */
interface IPlatform<Platform : Any, PlatformPlayer : Any, Player : ITabPlayer<PlatformPlayer>, PlaceholderApi: IPlaceholderApi<Player>> {
    /**
     * The platform object on the current platform.
     */
    val platform: Platform

    /**
     * The placeholder API implementation for this platform.
     */
    val placeholderApi: PlaceholderApi?

    /**
     * All the current online players in an immutable collection.
     */
    val onlinePlayers: Collection<Player>

    /**
     * Get an online player by their UUID.
     */
    fun getPlayer(uuid: UUID): Player?

    /**
     * Get an online player by their exact name.
     */
    fun getPlayer(name: String): Player?
}